### DISCUSSION TITLE
Switch behavior changes when MQTT in failed connection state.

### DISCUSSION THREAD

**Discussion by bdgit** [arendst/Tasmota#11508] [Q&A] [answered]
My garage door controller was built to control and monitor two doors using:
- ESP8266 NodeMCU
- Dual Relay Module (one relay to toggle each garage door respectively)
- Two typical window contacts (one to monitor the closed status of each door)

Originally, I wrote my own basic software, but decided to switch to Tasmota well over a year ago.

It worked/works great.  Although, a glitch was found, by chance.

Some time last summer, I decided to do away with my RaspberryPI openHABian server.  All the exact same services were recreated using multiple docker containers (on a small server).  Mosquitto was and still handles the MQTT needs.

During this process, the glitch was noticed.  While I explored issues with the hardware and new openHAB environment, it turned out to be something to do with Tasmota.  As my MQTT server is rarely if ever offline, I've put exploring this on hold for a while.  
- Today started on v8.5.0.
- I don't recall if I experienced with prior versions.
- Today I upgraded to 9.2, 9.3, and 9.3.1.2
- All versions had the same results (based on the tests I performed)

There is a chance my configuration could be improved, regardless, I feel like this glitch may be undesirable.

Door 1 is monitored by Switch3 and controlled using Relay_i1 (hint: odd numbers)
Door 2 is monitored by Switch4 and controlled using Relay_i2 (hint: even numbers)

Template:
{"NAME":"Garage","GPIO":[0,0,0,0,257,256,0,0,163,0,162,0,0,0],"FLAG":0,"BASE":18}

Rule1:
on power1#State=1 do backlog delay 1; power1 0 endon

Rule2:
on power2#State=1 do backlog delay 1; power2 0 endon

-----------------------------------------------------------------------
The expected results occur when connected to the MQTT server.

Test 1:
- Connected to WiFi and MQTT
- Starting:
  - door 2 fully CLOSED and not moving.
  - door 1 fully closed
- Push the directly connected physical button in the garage to OPEN door 2
- Console
  - MQT ... cmnd/POWER4 = ON
- The visual result is:
  - door 2 being fully open (due to the physically button and no interference from the controller)
  - door 1 being fully closed

Test 2:
- Connected to WiFi and MQTT
- Starting:
  - door 2 fully OPEN and not moving.
  - door 1 fully closed
- Push the directly connected physical button in the garage to CLOSE door number 2
- Console
  - MQT ... cmnd/POWER4 = OFF
- The visual result is:
  - door 2 being fully closed
  - door 1 being fully closed

-------------------------------------------------------------------------------
The unexpected results occur when NOT connected to the MQTT server.

Test 3:
- Connected to WiFi and  not-connected to MQTT
- Starting:
  - door 2 fully CLOSED and not moving.
  - door 1 fully closed
- Push the directly connected physical button in the garage to OPEN door 2
- Console
  - RSL ... stat/RESULT= {"POWER1":"ON"}
    (1st) triggered instead of **cmnd/power4:on**
    door 1 starts opening
  - RSL ... stat/POWER1 = ON
  - RUL ... POWER1#STATE=1 performs "backlog delay 1; power1 0"
    rule because of power1:on
  - RSL ... stat/RESULT = {"Delay":2}
  - RSL ... stat/RESULT = {"POWER1":"OFF"}
  - RSL ... stat/POWER1 = OFF
  - RSL ... stat/RESULT = {"POWER1":"ON"}
    (2nd) triggered instead of **power3:on** (due to power1:on happening)
    door 1 stops opening
  - RSL ... stat/POWER1 = ON
  - RUL ... POWER1#STATE=1 performs "backlog delay 1; power1 0"
    rule because of power1:on
  - RSL ... stat/RESULT = {"Delay":1}
  - RSL ... stat/RESULT = {"POWER1":"OFF"}
  - RSL ... stat/POWER1 = OFF
- It ends here because 2nd trigger of garage door stops it on the way up.  Switch3 is only changes on close/un-close
- This flow makes sense based on the 1st trigger and the cascading affect.
- The visual result is:
  - door 2 being fully open (due to the physically button and no interference from the controller)
  - door 1 being cracked a half foot (due to the sequence above)

Test 4:
- Connected to WiFi and  not-connected to MQTT
- Starting:
  - From Test 3 Result
- Push the directly connected physical button in the garage to CLOSE door 2
- Console
  - RSL ... stat/RESULT= {"POWER1":"OFF"}
    (1st) triggered instead of **cmnd/power4:off**
  - RSL ... stat/POWER1 = OFF
- The visual result is:
  - door 2 being fully closed (due to the physically button and no interference from the controller)
  - door 1 being cracked a half foot (due to result of Test 3)

-------------------------------------------------------------------------------
I ran various other tests, including pushing the physical button for door 1.  The results follow the same path.  With MQTT, POWER3 occurs properly.  Without MQTT, it starts with the 2nd trigger from test 3.

I did not test with WiFi disconnected.

I have not looked into the source code (probably way above my understanding and familiarity), but will make an attempt soon.

If this belongs as an issue, please let me know.  Thought I would start here.

Thank you for reading.

**Comment by sfromis** [comment, marked as answer]
While I've not parsed the details of what you report, by design Tasmota will in some cases react to MQTT unavailability by [falling back](https://tasmota.github.io/docs/Buttons-and-Switches/#switchtopic) to more basic behavior, when depending on MQTT could leave the device inoperable.

**Reply by bdgit** [reply]
Thank you!
Your suggestion put me on the right path.

Turns out I did have a custom SwitchTopic set.  I failed to record that in my notes.  That said, I was not aware of this fall back scenario.

Changes I performed (since I'm on version 9.1+):
- 17:16:07.759 CMD: SetOption114 1
- 17:16:07.766 MQT: garage/door/controller/stat/RESULT = {"SetOption114":"ON"}
- 17:16:33.532 CMD: SwitchTopic 0
- 17:16:33.541 MQT: garage/door/controller/stat/RESULT = {"SwitchTopic":""}

Tests with MQTT connected and without it connected were the same:
- Pushed physical button to OPEN door 2
  17:19:55.885 MQT: garage/door/controller/stat/RESULT = {"Switch4":{"Action":"ON"}}
- Pushed physical button to CLOSE door 2
  17:20:10.598 MQT: garage/door/controller/stat/RESULT = {"Switch4":{"Action":"OFF"}}

-------------

### A follow up question:
While I don't understand why the fall back happens; it is obviously known and potentially 'not optimal'... as in my case.  What I am curious about is:
Why would both Switch3 and Switch4 fall back to Power1?

**Reply by sfromis** [reply]
I've not been involved in reasoning behind the fall back, my impression is what I already stated, to mitigate the issue of devices becoming inoperable without MQTT available. It could make sense to, say, be able to switch a lamp on/off while the network is croaking.

I don't know your full config, but it sounds like you expect that assigning number 3 and 4 instead of 1 and 2 (which you are not using?) would make a difference. Coupling is basically not by numbers, but by sequence. In principle, you should start each type from number 1. However, as I'm far from well-versed in the intricacies of [Buttons and Switches](https://tasmota.github.io/docs/Buttons-and-Switches/), I'll not pretend to have all the answers.

**Reply by bdgit** [reply]
Totally get the desire for things to operate when MQTT and possibly WiFi are offline.  That is exactly the basis for my questions.  Unfortunately, the fallback results in a different result when MQTT is offline.  

Oh, I should note... external MQTT connected things (openHAB, etc...) have no influence when MQTT is online with it.

The sequential aspect you mentioned did challenge me once in the past on another device.  I briefly considered that here and (correctly or not dismissed).  That is what led me to the follow-up question.  If the sequential aspect was involved, I would have expected Switch3 to potentially be 1 and Switch4 to be 2.  Not both going to 1.

I started to work my way through the code.  Kind of looking for a default which falls to Power1.

Leaves me curious what would have happened if I had a Button1 for Relay1 plus a Button2 for Relay2.  How would the fallback.  A test to explore when I'm bored in the future.

Thank you again for pointing me toward a solution... and the discussion.
